### PR TITLE
[Feature] Edge softmax on a subset of edges in the graph.

### PR DESCRIPTION
**Pull request by yzh119**
## Description
Our original design of `edge_softmax` applies softmax on all edges of the graph. However, for some scenarios (e.g. beam search) users need to apply edge softmax on a subset of edges in the graph.

This PR changes the api from
```python
def edge_softmax(graph, logits)
```
to
```python
def edge_softmax(graph, logits, eids=ALL)
```

This change is compatible with previous API and allow users to specify the indices of edges to apply edge softmax on.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

## Changes
- dgl.nn.pytorch.softmax
- dgl.nn.mxnet.softmax

### CODE REVIEW

**Comment by yzh119** [comment]
@ylfdq1118  I've run the gat example on both PyTorch and MXNet, there is no memory leak.

**Comment by yzh119** [comment]
@ylfdq1118 I've updated the code with `edge_subgraph`.
